### DISCUSSION TITLE
TypeVar of optional type

### DISCUSSION THREAD

**Discussion by raceychan** [microsoft/pyright#6543] [Q&A] [answered]
```python

def dummy[OptInt: int | None](num: OptInt) -> OptInt:
    return None if num is None else num
```
pylance says:

Expression of type "int* | None" cannot be assigned to return type "OptInt@dummy"
  Type "int* | None" cannot be assigned to type "OptInt@dummy"
  
 is pylance coorect here? what should we do to get this right.
 
  Optional[int] is probably not what we want, since we want the function return None only when input is None, similar to the Maybe type in functional programming language.

**Comment by erictraut** [comment, marked as answer]
A generic function is not the right tool here. This is what `@overload` was designed to handle.

Code sample in [pyright playground](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMpgBuApiADZgCGAJgFC0ACRpFNt1xwU1ArhBHAAUKPgC5MKGAEooAWgB8EmKNpQ1UAHRbGzclTocuvfkJERxAOTApiMhVCs2V6zdsPc%2BA4WKVQAPg7WtnKKqPgBjsTO6iDEMDwgKIE2mFxmmADOycRQxGQZOWZAA)

```python
from typing import overload

@overload
def dummy(num: int) -> int:
    ...
@overload
def dummy(num: None) -> None:
    ...
def dummy(num: int | None) -> int | None:
    return None if num is None else num
```

**Reply by raceychan** [reply]
thanks eric, you have been very responsive and helpful, i saw this usage of typing.overload, but i could not think through how is 
this not allow, `"Type "int* | None" cannot be assigned to type "OptInt@dummy"`, but isn't OptInt@dummy supposed to be the same as `Type "int* | None' ?

**Reply by erictraut** [reply]
The answer to your question gets pretty deep into type theory ... probably deeper than you're interested in. The type `int*` means that it's an intersection between `int` and `OptInt`. The Python type system doesn't formally support intersections, but pyright implements them in a limited form under the covers to support certain cases like this. The type narrowing that results from the ternary operator is `int* | None`. If Python officially supported intersections, you'd see this type printed as `(int & OptInt) | None`. This type isn't type compatible with the return type `OptInt`.

As I said, `@overload` is the correct tool for the job here. That should solve your problem nicely.

**Reply by raceychan** [reply]
yes, @overload indeed solveed the problem, i saw one answer from stackoverflow suggesting typing.overload is the way to go, and then you confirmed it, but I just couldn't stop wondering why, it turned out that 'int* | None' is not the samething as 'int | None', TIL.

**Comment by Azureblade3808** [comment]
You may try
```python
def dummy[SomeInt: int](num: SomeInt | None) -> SomeInt | None:
    return None if num is None else num
```

**Reply by raceychan** [reply]
thanks
